### PR TITLE
feat(TextArea): add component

### PR DESCRIPTION
**Pull request by lucivpav**
Implements #755

|Prop|Meaning|
|---|---|
|accessibility?|Accessibility behavior if overridden by the user.|
|defaultValue?|The default value of the text area.|
|onChange?|Called on change.|
|value?|The value of the text area.|
|disabled?|The text area becomes read-only.|

[Semantic UI's TextArea for reference](https://react.semantic-ui.com/addons/text-area)
[Mozilla's docs on `<textarea>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea)

### CODE REVIEW

File: docs/src/components/ComponentPlayground/ComponentPlayground.tsx
Comment (lucivpav) [inline docs/src/components/ComponentPlayground/ComponentPlayground.tsx]: There is a problem (along the lines: "A component is changing an uncontrolled input of type text to be controlled error in ReactJS") when one plays with `placeholder` in the playground. So I decided to disable it.

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (layershifter) [inline packages/react/src/components/TextArea/TextArea.tsx]: There is no need in it as it will be passed down as HTML prop

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (layershifter) [inline packages/react/src/components/TextArea/TextArea.tsx]: https://github.com/stardust-ui/react/pull/1897/files#r322261384

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (layershifter) [inline packages/react/src/components/TextArea/TextArea.tsx:93]: nit, let's move it upper

File: docs/src/examples/components/TextArea/index.tsx
Comment (layershifter) [inline docs/src/examples/components/TextArea/index.tsx]:
```suggestion
  <>
```

File: packages/react/src/lib/accessibility/Behaviors/TextArea/textAreaBehavior.ts
Comment (layershifter) [inline packages/react/src/lib/accessibility/Behaviors/TextArea/textAreaBehavior.ts:6]: Let's add a description and tests for it

**Review by layershifter — COMMENTED** [review]
Let's add a common test

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (layershifter) [inline packages/react/src/components/TextArea/TextArea.tsx]:
```suggestion
export default withSafeTypeForAs<typeof TextArea, TextAreaProps, 'textarea'>(TextArea)
```

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (miroslavstastny) [inline packages/react/src/components/TextArea/TextArea.tsx]:
```suggestion
        className={classes.root}
```

For each slot you need to set `className` to value from `classes`.
For all slots other than `root`, you must concat that with slot class name using `cx`.

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (miroslavstastny) [inline packages/react/src/components/TextArea/TextArea.tsx]: Never pass `styles` to `ElementType`. `className` handles the styling.

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (miroslavstastny) [inline packages/react/src/components/TextArea/TextArea.tsx]:
```suggestion
        {...applyAccessibilityKeyHandlers(accessibility.keyHandlers.root, unhandledProps)}
```

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (miroslavstastny) [inline packages/react/src/components/TextArea/TextArea.tsx]: Remove this. You do not have any slots, just the `root` which is `TextArea.className`

**Comment by miroslavstastny** [comment]
> Are there any props you would change/add? Such as `minHeight?`?

No `minHeight` we have `design` prop for that. Once you set correct `className` on `ElementType` this will work: `<TextArea design={{minHeight: '400px'}} />`

**Comment by lucivpav** [comment]
> Let's add a common test

I added unit test as well as a behavior test

**Comment by lucivpav** [comment]
Bug: now the maxLength doesn't work. Need to use `maxlength` instead of `maxLength` prop.

**Comment by lucivpav** [comment]
There should be a text text to the textarea that has limited number of characters. (And used aria-describedby/labeledby)

**Comment by lucivpav** [comment]
Do not merge this without adding accessibility features with Milan. (Let Milan know once this PR is approved)

File: docs/src/examples/components/TextArea/Types/TextAreaDisabledExample.shorthand.tsx
Comment (layershifter) [inline docs/src/examples/components/TextArea/Types/TextAreaDisabledExample.shorthand.tsx]:
```suggestion
const TextAreaDisabledExample = () => <TextArea disabled placeholder="You can't type here" />
```

_nit_

File: docs/src/examples/components/TextArea/Types/index.tsx
Comment (layershifter) [inline docs/src/examples/components/TextArea/Types/index.tsx]: `Disabled` should be under `State`

File: docs/src/examples/components/TextArea/Types/index.tsx
Comment (layershifter) [inline docs/src/examples/components/TextArea/Types/index.tsx]: `Usage` section?

File: docs/src/examples/components/TextArea/Types/index.tsx
Comment (layershifter) [inline docs/src/examples/components/TextArea/Types/index.tsx:12]: Let's add an example for `value` with `knobs` under `State` section

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (layershifter) [inline packages/react/src/components/TextArea/TextArea.tsx]:
```suggestion
```

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (layershifter) [inline packages/react/src/components/TextArea/TextArea.tsx]:
```suggestion
```

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (layershifter) [inline packages/react/src/components/TextArea/TextArea.tsx]:
```suggestion
   /** The text area becomes read-only. */
```

File: packages/react/src/lib/accessibility/Behaviors/TextArea/textAreaBehavior.ts
Comment (layershifter) [inline packages/react/src/lib/accessibility/Behaviors/TextArea/textAreaBehavior.ts]:
```suggestion
* Adds attribute 'aria-disabled=true' based on the property 'disabled'.
```

File: packages/react/test/specs/behaviors/textAreaBehavior-test.tsx
Comment (layershifter) [inline packages/react/test/specs/behaviors/textAreaBehavior-test.tsx]: I think that this file is redundant as should be tested by comments:
https://github.com/stardust-ui/react/blob/master/.github/test-a-feature.md#behavior-tets

https://github.com/stardust-ui/react/pull/1897/files#r322706395

File: packages/react/test/specs/components/TextArea/TextArea-test.tsx
Comment (layershifter) [inline packages/react/test/specs/components/TextArea/TextArea-test.tsx]: Not sure, but `textarea` is `root` so `isConformant(TextArea)` should work without any config

File: packages/react/test/specs/components/TextArea/TextArea-test.tsx
Comment (layershifter) [inline packages/react/test/specs/components/TextArea/TextArea-test.tsx]:
```suggestion
  describe('conformance', () => {
```

`isConformant` creates own `describe`

File: packages/react/test/specs/components/TextArea/TextArea-test.tsx
Comment (layershifter) [inline packages/react/test/specs/components/TextArea/TextArea-test.tsx]:
```suggestion
      wrapper.simulate('change', { target: { value } })
```

Let's avoid useless helper function

File: packages/react/test/specs/components/TextArea/TextArea-test.tsx
Comment (layershifter) [inline packages/react/test/specs/components/TextArea/TextArea-test.tsx]:
```suggestion
      const wrapper= mount(<TextArea value={controlledTextAreaValue} />)
```

It's common to call it `wrapper`
https://airbnb.io/enzyme/docs/api/ReactWrapper/mount.html

File: packages/react/test/specs/components/TextArea/TextArea-test.tsx
Comment (layershifter) [inline packages/react/test/specs/components/TextArea/TextArea-test.tsx:31]: Duplicate of previous `describe`?

File: packages/react/test/specs/components/TextArea/TextArea-test.tsx
Comment (layershifter) [inline packages/react/test/specs/components/TextArea/TextArea-test.tsx]: General thing, in SUIR it was common to structure tests in this way and the same thing is covered in our docs:
https://github.com/stardust-ui/react/blob/master/.github/test-a-feature.md#writing-tests

```tsx
describe('TextArea', () => {
  isConformant(TextArea)

  describe('defaultValue', () => {
     test('sets "defaultValue" as initial "value"', () => {
       const wrapper = mount(<TextArea defaultValue="foo" />)
       
       expect(wrapper.prop('value')).toBe('foo')
     })
  })

  describe('onChange', () => {
    test('is called with (e, data)', () => {
      const onChange = jest.fn()
      const wrapper = mount(<TextArea onChange={onChange} />)

      wrapper.simulate('change', { target: { value: 'foo' })
      expect(onChange).toBeCalledTimes(1)
      expect(onChange).toBeCalledWith(expect.objectContaining({ type: 'change' }), expect.objectContaining({ value: 'foo' })
    })
  })
})
```

File: docs/src/components/ComponentPlayground/ComponentPlayground.tsx
Comment (layershifter) [reply docs/src/components/ComponentPlayground/ComponentPlayground.tsx]: Can be reenabled now?

**Review by kolaps33 — APPROVED** [review]

**Review by kolaps33 — APPROVED** [review]

File: docs/src/examples/components/TextArea/Types/index.tsx
Comment (lucivpav) [reply docs/src/examples/components/TextArea/Types/index.tsx]: You mean that I would use `useStringKnob()` for the value? Or other knobs for other props too?

File: packages/react/test/specs/components/TextArea/TextArea-test.tsx
Comment (lucivpav) [reply packages/react/test/specs/components/TextArea/TextArea-test.tsx]: I don't know why, but if I inline it, the test will fail

File: docs/src/components/ComponentPlayground/ComponentPlayground.tsx
Comment (lucivpav) [reply docs/src/components/ComponentPlayground/ComponentPlayground.tsx]: Since there is no longer `placeholder` prop on (nor `maxLength` prop), it can be enabled. But I am not sure about its value

**Comment by lucivpav** [comment]
@layershifter one more review please :)

File: docs/src/examples/components/TextArea/Usage/TextAreaMaxLengthExample.shorthand.tsx
Comment (layershifter) [inline docs/src/examples/components/TextArea/Usage/TextAreaMaxLengthExample.shorthand.tsx]:
```suggestion
      placeholder='No more than 5 letters can be here...'
```

Let's inline it

File: docs/src/examples/components/TextArea/Usage/TextAreaMaxLengthExample.shorthand.tsx
Comment (layershifter) [inline docs/src/examples/components/TextArea/Usage/TextAreaMaxLengthExample.shorthand.tsx]:
```suggestion
      maxLength={5}
```

File: docs/src/examples/components/TextArea/Usage/index.tsx
Comment (layershifter) [inline docs/src/examples/components/TextArea/Usage/index.tsx:6]:
```suggestion
  <ExampleSection title="Usage">
```

File: packages/react/test/specs/components/TextArea/TextArea-test.tsx
Comment (layershifter) [inline packages/react/test/specs/components/TextArea/TextArea-test.tsx]:
```suggestion
      const wrapper = mount(<TextArea defaultValue='foo' />)
```

Can we inline it? It's more obvious in the most cases

File: packages/react/test/specs/components/TextArea/TextArea-test.tsx
Comment (layershifter) [inline packages/react/test/specs/components/TextArea/TextArea-test.tsx]: Note sure that it's useful at all now as it tests the same thing as `AutoControlledComponent` does...

File: docs/src/examples/components/TextArea/Usage/index.tsx
Comment (lucivpav) [reply docs/src/examples/components/TextArea/Usage/index.tsx:6]: Actually, we don't specify `Usage`. As all the examples under Usage naturally belong to `Usage`.

**Review by layershifter — COMMENTED** [review]
Left minor comments, LGTM 👍 

Let's add some dumb styles and variables (for `margin: 0`, `resize: vertical`) to create boilerplate under Teams theme. And add an entry to `CHANGELOG.md` before merging 😼

File: packages/react/test/specs/components/TextArea/TextArea-test.tsx
Comment (lucivpav) [reply packages/react/test/specs/components/TextArea/TextArea-test.tsx]: But it's also more prone to errors, if you have to specify the same text at several places, you can make a typo. In the worst case, a test would pass while component is broken.

File: packages/react/test/specs/components/TextArea/TextArea-test.tsx
Comment (lucivpav) [reply packages/react/test/specs/components/TextArea/TextArea-test.tsx]: Does it mean the test code of `Input` is also irrelevant? https://github.com/stardust-ui/react/blob/master/packages/react/test/specs/components/Input/Input-test.tsx#L72

Comment (layershifter) [reply packages/react/test/specs/components/TextArea/TextArea-test.tsx]: Yep, it's strange

File: packages/react/test/specs/components/TextArea/TextArea-test.tsx
Comment (lucivpav) [reply packages/react/test/specs/components/TextArea/TextArea-test.tsx]: Ok, I replaced all the occurences of `testValue` by `foo`.

Comment (layershifter) [reply packages/react/test/specs/components/TextArea/TextArea-test.tsx]: At the same time it makes code less readable.

> But it's also more prone to errors, if you have to specify the same text at several places, you can make a typo.

It _should_ be the same only in single test, it's good to use [faker](https://www.npmjs.com/package/faker), actually.

File: docs/src/examples/components/TextArea/Usage/index.tsx
Comment (layershifter) [reply docs/src/examples/components/TextArea/Usage/index.tsx:6]:
```suggestion
  <ExampleSection>
```

So it should be or even

```suggestion
  <UsageSection>
```

Otherwise it extremely confusing

File: docs/src/examples/components/TextArea/Usage/TextAreaMaxLengthExample.shorthand.tsx
Comment (layershifter) [inline docs/src/examples/components/TextArea/Usage/TextAreaMaxLengthExample.shorthand.tsx]: Oops, `Fragment` is useless

```suggestion
```

Comment (lucivpav) [reply docs/src/examples/components/TextArea/Usage/TextAreaMaxLengthExample.shorthand.tsx]: Nice catch

File: packages/react/test/specs/components/TextArea/TextArea-test.tsx
Comment (lucivpav) [reply packages/react/test/specs/components/TextArea/TextArea-test.tsx]: I am using `faker` now. Hopefully, the code is still readable 🤔

File: packages/react/test/specs/components/TextArea/TextArea-test.tsx
Comment (lucivpav) [reply packages/react/test/specs/components/TextArea/TextArea-test.tsx]: Ok, I removed it from the other components, #1917

File: docs/src/examples/components/TextArea/Usage/index.tsx
Comment (lucivpav) [reply docs/src/examples/components/TextArea/Usage/index.tsx:6]: I agree its confusing 👍

Comment (lucivpav) [reply docs/src/examples/components/TextArea/Usage/index.tsx:6]: Let me create and use `UsageSection` in a seperate PR

File: docs/src/examples/components/TextArea/Usage/TextAreaMaxLengthExample.shorthand.tsx
Comment (layershifter) [inline docs/src/examples/components/TextArea/Usage/TextAreaMaxLengthExample.shorthand.tsx]:
```suggestion
  <TextArea placeholder='No more than 5 letters can be here...' maxLength={5} />
```

_nit_, should be handled in future by ESLint

**Review by layershifter — APPROVED** [review]

**Pull request merged**
